config/v1/types_infrastructure.go

```diff
@@ -505,9 +505,113 @@ type BareMetalPlatformStatus struct {
+
+	// ip is the IP address of the peer. It may be either IPv4 or IPv6.
+	//
+	// +kubebuilder:validation:Required
+	IP string `json:"ip"`
+}
+
+// OpenStackAPIBGPSpeaker species
+type OpenStackAPIBGPSpeaker struct {
+	// failureDomain is the name of a failure domain which this BGP
+	// configuration applies to.
+	//
+	// +kubebuilder:validation:Required
+	FailureDomain string `json:"failureDomain"`
+
+	// asn specifies the ASN to be used by the BGP speaker.
+	//
+	// +kubebuilder:validation:Required
+	ASN string `json:"asn"`
+
+	// peers is a list of all BGP peers of this speaker.
+	//
+	// +kubebuilder:validation:MinItems:=1
+	Peers []OpenStackAPIBGPPeer `json:"peers"`
+}
+
+type OpenStackAPIBGPConfiguration struct {
+	// apiVIPs is a list of IP addresses which will be published via BGP for
+	// use by the API server. It may contain both IPv4 and IPv6 addresses.
+	//
+	// +kubebuilder:validation:MinItems:=1
+	APIVIPs []string `json:"apiVIPs"`
+
+	// speakers is a list of BGP speaker configurations. We require a
+	// speaker configuration for every failure domain hosting a control
+	// plane node.
+	//
+	// +kubebuilder:validation:MinItems:=1
+	// +listType=map
+	// +listMapKey=failureDomain
+	Speakers []OpenStackAPIBGPSpeaker `json:"speakers"`
+}
+
+// OpenStackFailureDomain specifies a failure domain for an OpenStack server.
+// Specifically, it specifies a set of values which will be set on all machines
+// using the failure domain.
+type OpenStackFailureDomain struct {
+	// name is the unique name of this failure domain.
+	//
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+
+	// computeZone specifies the OpenStack Compute availability zone for
+	// all servers in this failure domain.
+	//
+	// If not specified, the servers are provisioned in the default Compute
+	// availabity zone.
+	// +kubebuilder:validation:Optional
+	ComputeZone string `json:"computeZone,omitempty"`
+
+	// storageZone specifies the OpenStack storage availability zone for
+	// all volumes in this failure domain.
+	//
+	// If not specified, the volumes are provisioned in the default storage
+	// availabity zone.
+	// +kubebuilder:validation:Optional
+	StorageZone string `json:"storageZone,omitempty"`
+
+	// subnetID specifies an OpenStack subnet ID which will be attached as
+	// the first NIC of every server in this failure domain.
+	//
+	// +kubebuilder:validation:Optional
+	SubnetID string `json:"subnetID,omitempty"`
+}
+
 // OpenStackPlatformSpec holds the desired state of the OpenStack infrastructure provider.
 // This only includes fields that can be modified in the cluster.
-type OpenStackPlatformSpec struct{}
+type OpenStackPlatformSpec struct {
+	// failureDomains is a list of failure domains available to Machines.
+	// Values specified in a failure domain will be set on Machines in that
+	// failure domain.
+	//
+	// +listType=map
+	// +listMapKey=name
+	FailureDomains []OpenStackFailureDomain `json:"failureDomains,omitempty"`
```

Comment:
Is this optional?

Reply:
Yes.

Possibly related: is an empty list of FailureDomains different to no list of FailureDomains? 🤔 i.e. `[]OpenStackFailureDomain` vs `*[]OpenStackFailureDomain`. I don't think an empty list would have a useful meaning: explicitly don't schedule these servers in any FailureDomain? If we're confident that's the case I prefer the simpler type here. (This came up recently with the upstream API.)

Reply:
Add an `// +optional` then please.

Using a reference is preferred in configuration APIs unless you absolutely need to have a distinction between unset and emtpy, which I don't think you do in this case?

Reply:
added